File: sdks/ruby/lib/solvapay/results.rb
# frozen_string_literal: true

module SolvaPay
  class PayablePaywallResult
    attr_reader :kind, :content

    def initialize(content:)
      @kind = "paywall"
      @content = content.freeze
      freeze
    end
  end

  class PayableAllowResult
    attr_reader :kind, :customer_ref, :decision

    def initialize(customer_ref:, decision:, track_success:, track_fail:)
      @kind = "allow"
      @customer_ref = customer_ref
      @decision = decision.freeze
      @track_success = track_success
      @track_fail = track_fail
    end

    def track_success(duration: nil, metadata: nil)
      @track_success.call(duration: duration, metadata: metadata)
    end

    def track_fail(error, duration: nil, metadata: nil)
      @track_fail.call(error, duration: duration, metadata: metadata)
    end
  end
end

File: sdks/python/python/solvapay/results.py
"""Gate result types for the idiomatic Python facade (Step 41-d / §2.4)."""

from __future__ import annotations

from collections.abc import Callable, Mapping
from dataclasses import dataclass
from typing import Literal


@dataclass(frozen=True)
class PayablePaywallResult:
    """Customer is gated — callers format a 402 from ``content``."""

    kind: Literal["paywall"]
    content: Mapping[str, object]


@dataclass
class PayableAllowResult:
    """Customer is allowed — record usage via ``track_success`` / ``track_fail``."""

    kind: Literal["allow"]
    customer_ref: str
    decision: Mapping[str, object]
    track_success: Callable[..., None]
    track_fail: Callable[..., None]


PayableGateResult = PayablePaywallResult | PayableAllowResult

__all__ = [
    "PayableAllowResult",
    "PayableGateResult",
    "PayablePaywallResult",
]
